### PR TITLE
Graph Store documentation

### PR DESCRIPTION
**Pull request by ynx0**
This PR adds documentation for the Graph Store gall agent. I've edited and revised it to the best of my ability, but it may still need further modification. It could be a good idea to have a dev review the content for accuracy as well.

**General areas to review**
- Spelling/punctuation/grammar/formatting
  - clickable code links don't look clickable with the current formatting
- See if any inaccurate terminology is used
- Images don't properly render, may need to be uploaded to media.urbit.org or something
- See if headings are correct
- Old reference appears to be out of date and could use some updating
- Old reference needs to be moved to a better place (the one in userspace/landscape)
- I had to invent a decent amount of terminology which you may want to review for appropriateness

**Document-specific concerns**
- **Overview**
  - Need verification that the explanations for permissions (specifically %yes %no and %self) are accurate
- **Sur File Walkthrough**
  - Phrase "CRUD actions" could be improved
- **Validator Walkthrough**
  - The sentences similar to, "Return a permissions noun defined as: [admin: %yes, writer: %yes, reader: %no]" don't look right and could be improved

**TODO**

- [x] Images render properly
- [x] Document content is reviewed for accuracy
- [ ] Document is fully proofread
- [ ] Old graph store reference is refreshed and integrated with new docs structure

### CODE REVIEW

**Comment by ynx0** [comment]
cc @timlucmiptev

File: userspace/graph-store/advanced-info.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+
+### What happens when you add or remove a node?
+
+When adding nodes, Graph Store takes in a flat list of nodes that have nodes whose `index`es are arbitrarily deep. It does not allow for a node which has non-existent ancestors to be added (i.e. it doesn't silently create the node). Thus, it assumes every node up until the 2nd to last index is created (meaning it needs every ancestor to exist execpt it's parent), either already existing or included in the `graph-update`. To ensure consistent behavior, all nodes are added shallowest-index first, which ensures that no child is added before it's parent (if it exists). This is why any non-leaf node must have its parent exist, either already within the graph or in the update. However, children do not have to be nested within the parent's data structure in the update. They simply have to exist somewhere in the update, even at the top level is fine. Graph Store merges the flat list of nodes from an `%add-nodes` update into the fully connected graph structure.
```
Comment (tacryt-socryp) [inline userspace/graph-store/advanced-info.md:25]: Graph store takes in a (map index node) when adding nodes, not a flat list.

File: userspace/graph-store/advanced-info.md [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+Another constraint to be aware of is that in a map of indexed nodes: for each node, the index in map entry has to be the same index as the in that `node`'s post index.
+
+To see how Graph Store handles add-nodes, take a look at [`app/graph-store.hoon#L370-L417`](https://github.com/urbit/urbit/blob/e2ad6e3e9219c8bfad62f27f05c7cac94c9effa8/pkg/arvo/app/graph-store.hoon#L370-L417).
+
+
+### Permissions internals
+Internally, the permissions part of a mark (the grow arm) are built and then watches for changes by Graph Store. See [`app/graph-push-hook.hoon#L154`](https://github.com/urbit/urbit/blob/ac096d85ae847fcfe8786b51039c92c69abc006e/pkg/arvo/app/graph-push-hook.hoon#L154) for more info.
```
Comment (tacryt-socryp) [inline userspace/graph-store/advanced-info.md:33]: *watched, not watches

File: userspace/graph-store/overview.md [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+
+## Intro
+Graph store is a non-relational database suitable for use in building social media applications. It is primarily designed for storing text-based content and data that is generally threaded and nested. It is not suitable as a general purpose database for highly structured or pure binary data.
+
+<p align="center">
+  <img src="images/image1.png"/>
+</p>
+
+Graph Store is mainly in charge of two things: data storage and retrieval, and schema validation. There are two related tools: Graph Push Hook, which provides permissioning support to Graph Store and acts as a proxy layer to Graph Store for outside ships to access, and Graph Pull Hook, which can be used to request Graph Store data from other ships. Importantly, Graph Store itself does not have a built-in permissions system (or rather, is permissions-agnostic), and acts in a purely trusted manner, and assumes that all inputs are trusted This is why Graph Push Hook exists: to mediate untrusted requests from outside ships to your instance of Graph Store, rejecting invalid ones where a ship should not be able to modify the data.
```
Comment (tacryt-socryp) [inline userspace/graph-store/overview.md:17]: "trusted" needs period afterwards.

Comment (tacryt-socryp) [inline userspace/graph-store/overview.md:17]: After "purely trusted" maybe we can say something about Graph Store *only allowing local connections to talk to it* and not allowing foreign ships at all.

File: userspace/graph-store/sur-file-walkthrough.md [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+      time-sent=time
+      contents=(list content)
+  ==
+::
++$  content
+  $%  [%text text=cord]
+      [%mention =ship]
+      [%url url=cord]
+      [%code expression=cord output=(list tank)]
+      [%reference =uid]
```
Comment (tacryt-socryp) [inline userspace/graph-store/sur-file-walkthrough.md:46]: Out of date

File: userspace/graph-store/sur-file-walkthrough.md [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+These types are used to cryptographically sign a given post, so that the host of some content cannot act as an imposter, and post content impersonating as someone else. `signature` represents a triple of signed message of hash, author, and author's life at time of posting. These can be used to cryptographically attest to a message. The implementation is a form of asymmetric/public-key encryption, where `q` and `r` are data necessary to look up a ship's public key on azimuth, which can be used to verify the validity of the message.
+
+### Content Types
+```
++$  content
+  $%  [%text text=cord]
+      [%mention =ship]
+      [%url url=cord]
+      [%code expression=cord output=(list tank)]
+      [%reference =uid]
```
Comment (tacryt-socryp) [inline userspace/graph-store/sur-file-walkthrough.md:83]: Out of date

File: userspace/graph-store/sur-file-walkthrough.md [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+--
+```
+
+`content` enumerates all the possible content types that a post can have. The possible content types can be:
+
+- **Text** - representing plain text content
+- **Url** - specific data type for urls
+- **Mention** - mentioning another ship
+- **Code** - a pair of a piece of code that was executed and its result (static data, no execution takes place inside of Graph Store)
+- **Reference** - a reference to another post. Uses the `uid` type under the hood
```
Comment (tacryt-socryp) [inline userspace/graph-store/sur-file-walkthrough.md:96]: Out of date

File: userspace/graph-store/validator-walkthrough.md [hunk 7]
```diff
@@ -0,0 +1,485 @@
...
+
+#### Schema
+
+<p align="center">
+  <img src="images/image12.png"/>
+</p>
+
+Here, a notebook, which is a collection of blog posts (called notes), is represented by the root graph. All data associated with the blog post is represented by the top level node, which is the note itself along with the associated comments. One level deeper, we see two container structures. The first one is the post revision container; it holds the edit history of your blog post. Every child node of this corresponds to the actual title and text of your blog post. The second one is the comments container. This represents the comment section of your blog post. Every child node of this is not a comment, but a comment revision container, which, as before, contains the edit history of your comment.
+
+Here's its validator
```
Comment (tacryt-socryp) [inline userspace/graph-store/validator-walkthrough.md:289]: Please include links to the actual validator files in question as well.

File: userspace/graph-store/sur-file-walkthrough.md (hunk 4 above)
Comment (ynx0) [reply userspace/graph-store/sur-file-walkthrough.md:46]: Question about the `reference` type: if there is an unmanaged graph (i.e. one that is not tied to a group iirc), then how do references work for that? Cause a `%graph` reference needs a rid to a group. Do references only work for managed graphs?

File: userspace/graph-store/sur-file-walkthrough.md [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+      [%unarchive-graph =resource]
+      [%run-updates =resource =update-log]
+    ::
+    ::  NOTE: cannot be sent as pokes
+    ::
+      [%keys =resources]
+      [%tags tags=(set term)]
+      [%tag-queries =tag-queries]
+  ==
+--
```
Comment (ynx0) [inline userspace/graph-store/sur-file-walkthrough.md:207]: Just realized, this section is out of data based on the latest breaking changes as well. I have to figure out what the new stuff is.

File: userspace/graph-store/sur-file-walkthrough.md (hunk 4 above)
Comment (tacryt-socryp) [reply userspace/graph-store/sur-file-walkthrough.md]: Even unmanaged graphs have a group associated (but it's an unmanaged / hidden group). The group in that case has the same resource id as the graph.

**Review by tacryt-socryp — APPROVED** [review]
LGTM, thank you for your service

**Comment by ynx0** [comment]
@tacryt-socryp as a result of our discussion a while ago, I decided to elide the explanations for permissioning. I've stubbed it out as a TODO temporarily, but ideally we'd like to bring them back once permissions are fixed. What would you recommend we do in the long-run? Is there any timeline for the changes to the permissioning scheme?

**Comment by tacryt-socryp** [comment]
@jalehman If you're happy with the formatting, I say we go ahead and merge this.

**Review by jalehman — CHANGES_REQUESTED** [review]
The images aren't coming through—they need to be hosted on media.urbit.org. @ynx0 , if you could 

1. Rename them all with more meaningful names (so they're easy to differentiate) 
2. Change all of the links to point to https://media.urbit.org/docs/userspace/graph-store/<new image name>
3. I can get them uploaded to our media bucket at that location. 

The backticks in "`sur` file overview" don't render properly. Let's change the name of that section to something like "Data Structure Overview" or something.

File: userspace/graph-store/overview.md [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+  <img src="https://media.urbit.org/docs/userspace/graph-store/1_graph_store_overview.png"/>
+</p>
+
+Graph Store is mainly in charge of two things: data storage and retrieval, and schema validation. There are two related tools: Graph Push Hook, which provides permissioning support to Graph Store and acts as a proxy layer to Graph Store for outside ships to access, and Graph Pull Hook, which can be used to request Graph Store data from other ships. Importantly, Graph Store itself does not have a built-in permissions system (or rather, is permissions-agnostic), and acts in a purely trusted manner, and assumes that all inputs are trusted. However, it only allows local, authenticated connections to communicate with it, meaning that by default, no one else on the network may modify the state of your Graph Store agent. This is why Graph Push Hook exists: to mediate untrusted requests from outside ships to your instance of Graph Store, rejecting invalid ones where a ship should not be able to modify the data.
+
+## Data Structures
+### Posts
+
+<p align="center">
+  <img src="https://media.urbit.org/docs/userspace/graph-store/2_post_explanation_diagram.png"/>
```
Comment (jalehman) [inline userspace/graph-store/overview.md:23]:
```suggestion
  <img src="https://media.urbit.org/docs/userspace/graph-store/2_post_diagram.png"/>
```

**Review by jalehman — APPROVED** [review]
This looks good to go—just delete the images from this PR since they're now pushed up to the bucket, and we'll get this merged!

**Pull request merged**